src/Microsoft.SqlTools.ServiceLayer/Connection/ConnectionService.cs

```diff
@@ -345,17 +413,106 @@ public bool Disconnect(DisconnectParams disconnectParams)
                 return false;
             }
 
-            // Cancel if we are in the middle of connecting
-            if (CancelConnect(new CancelConnectParams() { OwnerUri = disconnectParams.OwnerUri }))
+            ConnectionInfo info = null;
+
+            // If we are disconnecting all existing connections for the provided URI
+            if (disconnectParams.Type == null)
             {
-                return false;
-            }
+                // Cancel all pending connections
+                foreach (KeyValuePair<Tuple<string, ConnectionType>, CancellationTokenSource> entry in CancelTupleToCancellationTokenSourceMap)
+                {
+                    string entryConnectionUri = entry.Key.Item1;
+                    ConnectionType entryConnectionType = entry.Key.Item2;
+                    if (disconnectParams.OwnerUri.Equals(entryConnectionUri))
+                    {
+                        CancelConnect(new CancelConnectParams() { OwnerUri = disconnectParams.OwnerUri, Type = entryConnectionType });
+                    }
+                }
 
-            // Lookup the connection owned by the URI
-            ConnectionInfo info;
-            if (!ownerToConnectionMap.TryGetValue(disconnectParams.OwnerUri, out info))
+                // Lookup the connection owned by the URI
+                if (!ownerToConnectionMap.TryGetValue(disconnectParams.OwnerUri, out info))
+                {
+                    return false;
+                }
+
+                // Close all open connections
+                bool closeFailed = false;
+                List<ConnectionType> connectionsToRemove = new List<ConnectionType>();
+
+                foreach (KeyValuePair<ConnectionType, DbConnection> entry in info.ConnectionTypeToConnectionMap)
+                {
+                    try
+                    { 
+                        entry.Value.Close();
+                        connectionsToRemove.Add(entry.Key);
+                    }
+                    catch (System.Data.Common.DbException e)
+                    {
+                        closeFailed = true;
+                    }
+                }
+
+                // If there we no errors calling Close(), simply clear the map and continue
+                if (!closeFailed)
```

Comment:
If an exception happened, it more likely means something is wrong with the connection, not that it'll still be usable. I'd recommend removing all connections either way - it's the stated intent and there's not much we can do if close errors. The connection's still broken and the best thing to do is clear up